By
### Bernardo Vimpi
April 2020

### PROJECT OBJECTIVE AND SCOPE

In this project I use feature selection through Standardazation and employ sklearn to evaluate liniear regression and predict the GPA scores of a given student.

###### FEATURE SELECTION VIA STANDARDIZATION

### Import the relevant libraries

In [167]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LinearRegression


### Data Load

In [168]:
data=pd.read_csv('original.csv')
data.head()

SAT  Rand 1,2,3   GPA
0  1714           1  2.40
1  1664           3  2.52
2  1760           3  2.54
3  1685           3  2.74
4  1693           2  2.83

#### DESCRIPTIVE STATISTICS OF THE DATA 

In [169]:
data.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

### Multiple Regression

#### Dependent and Independent Variables

In [170]:
x=data[['SAT', 'Rand 1,2,3']]
y=data['GPA']

#### STANDARDIZATION

In [171]:
#Lets import stardard scaler
from sklearn.preprocessing import StandardScaler


In [172]:
scaler=StandardScaler()

In [173]:
scaler.fit(x)

StandardScaler(copy=True, with_mean=True, with_std=True)

#### DATA TRANSFORMATION USING SCALER

In [174]:
x_scaled=scaler.transform(x)

In [175]:
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

#### Regression using Scaled Features

In [176]:
reg=LinearRegression()
reg.fit(x_scaled, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [177]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [178]:
reg.intercept_

3.330238095238095

### SUMMARY TABLE

In [179]:
reg_summary = pd.DataFrame([['Bias'],['SAT'], ['Rand 1,2,3']], columns = ['Features'])
reg_summary['Weights'] = reg.intercept_,reg.coef_[0], reg.coef_[1]
reg_summary

Features   Weights
0        Bias  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

#### Results and Analysis

'Weights' is the Machine learn word for coeficient.The name comes from the fact that the bigger the weights the bigger the inpact of the regression. It carries weight on the results. Intercept is also sometimes used in place of the Bias. Thats why I have included it in here.

Interpretation: The closer the Weight is to zero, the smaller its impact and the bigger the weight, the bigger its impact. In our case, the weight of SAT is 0.1718 which is much bigger than the weight of Rand 1,2,3 which is -0.0070. This leads us through feature selections via standardization.

It is clear than Rand 1,2,3 barely has any contribution to our output. It will make little to no much difference if we remove it from the model or if we let it stay there. 

### PREDICTIONS USING THE STANDARDIZED MODEL

In [180]:
#the numbers 2 and 1 are randmly assigned to each of the students.

new_data=pd.DataFrame(data=[[1700,2], [1800,1]], columns=['SAT', 'Rand 1,2,3'])
new_data


SAT  Rand 1,2,3
0  1700           2
1  1800           1

In [181]:
reg.predict(new_data)

array([295.39979563, 312.58821497])

##### Results and Analysis: 
These results are confusing since they are not even related to GPA. The reason is because the regression model was trained on standardized inputs.Therefore, it expects values that are of the same magnitude as in the Train process (Train input(x_scaled). The new data frame news t be standardized in the same was as it was arranged in the input x with the same mean and standard deviation.

### New Methodology

In [182]:
new_data_scaled = scaler.transform(new_data)
new_data_scaled

array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

#### Results and Analysis
The results of the new methodology is display in the above array which also is reflective of the Trained Model (Train inputs(x_scaled)).

#### USING THE PREDICT DATA FRAME

In [183]:
reg.predict(new_data_scaled)

array([3.09051403, 3.26413803])

### Conclusion 

The first student is predicted to have a GPA of 3.09 while the second a GPA of 3.26. To add into our analysis. What would happen if the Rand, 1,2,3 feature is removed? The theoretical analysis of this suggests that nothing should change. Well, lets find out next. 

#### REMOVING THE 'Rand 1,2,3' variable

### Create a new Simple Linear Regression

In [184]:
reg_simple = LinearRegression()
x_simple_matrix=x_scaled[:,0].reshape(-1,1)
reg_simple.fit(x_simple_matrix,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [185]:
reg_simple.predict(new_data_scaled[:,0].reshape(-1,1))

array([3.08970998, 3.25527879])

#### Analysis and Conclusion:
In comparison to the results from the New Methodology Data Frame above, we see the GPA results are slightly different. However, if we round up to 2 digits after the '.' we get exactly the same predicted GPA values of 3.09 and 3.26. The important lesson from this project is that while using sklearn, the results of the P-Values are not that highly relevant. Also, very often the final results are not that much affected if we decide to keep or drop off the insignificant features. The Weights will be so close to 0 that it will barely have any influnce on the predictions.

Thank you 

#### Bernardo Vimpi

April 2020